frontend/lib/resolution/resolution-queries.cpp

```diff
@@ -2835,31 +2874,52 @@ gatherAndFilterCandidatesForwarding(Context* context,
 
   // Resolve the forwarding expression's types & decide if we
   // want to consider forwarding.
-  const ResolvedFields* forwards = nullptr;
+  ResolvedFields forwards;
   UniqueString name = ci.name();
   if (name == USTR("init") || name == USTR("init=") || name == USTR("deinit")) {
     // these are exempt from forwarding
   } else if (auto ct = receiverType->getCompositeType()) {
-    auto useDefaults = DefaultsPolicy::USE_DEFAULTS;
-    const ResolvedFields& fields = fieldsForTypeDecl(context, ct,
-                                                     useDefaults);
-    if (fields.numForwards() > 0) {
-      // and check for cycles
-      bool cycleFound = emitErrorForForwardingCycles(context, ct);
-      if (cycleFound == false) {
-        forwards = &fields;
+    // Possible recursion here when resolving a function call in a forwarding
+    // statement:
+    //     record R { forwarding foo(); }
+    // We need to try resolving 'foo()' as a method on 'R', which eventually
+    // leads us back to this path here.
+    //
+    // By skipping the 'resolveForwardingExprs' check below, we effectively
+    // prevent forwarding statements from containing expressions that
+    // themselves require forwarding. For example, if you had a couple of
+    // forwarding statements like:
+    //     forwarding b;
+    //     forwarding bar();
+    // The 'isQueryRunning' check below would prevent resolving a method
+    // 'bar()' on 'b'.
```

Comment:
I think that resolving `bar()` to another forwarding statement would be confusing, so this seems like good behavior. It's probably worth clarifying in the spec.